### ISSUE TITLE
Directional forecasting 

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#745]
Contributors: @RNKuhns, @fkiraly, @mloning

### Estimators
* new estimator scitype: DirectionalForecaster
* reducer from Forecaster to DirectionalForecaster via time-series/tabular classification
* reducer from Forecaster to DirectionalForecaster via output binning
* override score in the reduced forecasters with the "right" binned losses!

```python
y_train, y_test = temporal_train_test_split(y)
forecaster = ARIMA()

# reduce forecaster to directional forecaster
dforecaster = Reducer(forecaster, bins=[...])
dforecaster.fit(y_train)
y_pred = dforecaster.predict(fh)

# score applies binning internally
dforecaster.score(y_test, fh)
```

### Losses 
* directional accuracy metrics are new metrics scitype - compare actual with binned; `__call__` method without parameters, class hyperparams are binning

```python
mean_directional_loss(y_test, y_pred, bins=[...])
```

Franz opinion (because parameter arguments should not be part of the scitype-specific call signatures, but go to the constructor, according to "universal interface spec")
```python
scoring = MeanDirectionalLoss(bins=[...])
scoring(y_test, y_pred)
```

wrapper for scikit-learn classification losses
```python
from sklearn import accuracy_loss
my_DL = make_directional_loss(accuracy_loss, bins=[...])
```

**Comment by Ayushjain9501** [comment]
Hi,
I am interested in contributing here. Could you guide on how to get started with this issue?

**Comment by Ayushjain9501** [comment]
I understand the implementation part, we check at every timestep whether the direction of change (increased, decreased) is in agreement with what actually happened.
However, I need some clarification on the inputs.
<ul>
<li>y_test : Is it a series with values like ['U', 'D', 'D'] or [340, 318, 362..]?
<li>y_pred : Same question 
</ul>
Could you please give some clarification on `bins` variable

**Comment by mloning** [comment]
If I remember correctly, we agreed to a assume `y_test` to be the real-valued time series (e.g. [340, 318, 362..]) and `y_pred` to be categories (e.g. ['U', 'D', 'D']).  

Is that right @fkiraly, @RNKuhns?

**Comment by fkiraly** [comment]
yes, I believe so

**Comment by fkiraly** [comment]
but my opinion is that `mean_directional_loss` should be a *class*, not a *function*, because it has parameters :smiley:

**Comment by mloning** [comment]
Yes we currently provide both functions and classes, where the class usually calls the function in `__call__` and keeps additional informations (e.g. bins or whether a greater value is better or worse).

**Comment by tabhishek432** [comment]
Hi, is anyone currently working on this issue? If not, can I work on this issue?

**Comment by mloning** [comment]
Hi @tabhishek432, this is a more advanced issue, if you're confident you can tackle it, you're welcome to work on it. If you first want to become a bit more familiar with sktime, perhaps this one is a good starting point: https://github.com/alan-turing-institute/sktime/issues/932

**Comment by RNKuhns** [comment]
@mloning and @fkiraly don't want to make things more confusing, but my memory is fuzzy on our discussion. 

About the assumption that the input `y_pred` would be real valued or categorical. 

The more I'm thinking about it, having `y_pred` be real valued at input makes more sense because real valued `y_pred` is needed (along with bins) to calculate the "increase" or "decrease" (or more generally the binned categories, but up/down is easier to talk about so I use that below too). 

Whether or not the prediction is an "increase" or "decrease" is generally determined by looking at `y_pred[t] - y_true[t-fh]` where `fh` is the horizon of interest. Note that for the first `fh` values in the dataset, you'd also need to be looking back into `y_train` to determine whether an increase or decrease occurred (but passing y_train is straight-forward now that we've merged #858). 

I've also got 2 clarification questions on the forecaster side of things: 
In the initial post we mention:

> - reducer from Forecaster to DirectionalForecaster via time-series/tabular classification
> - reducer from Forecaster to DirectionalForecaster via output binning

Was the intention for 2 new (non-public) classes (one for each bullet above)? Then a `make_directional_forecaster(..., type ='binned')` or `make_directional_forecaster(..., type='tsc')` for the actual reduction?

Also, for the `DirectionalForecaster` do we want predict to return a directional forecast or do we want that to return real values and we could implement the directional prediction via something a method like `predict_direction()`? Idea would be to make it kind of like how scikit-learn has `predict_proba()`. I guess the opposite could be true too -- `predict()` could predict the directional class in `DirectionalForecaster` but we could include a `predict_real` to predict real values in the output binning reduction case (real values wouldn't make sense in the time-series classification reduction case).

**Comment by fkiraly** [comment]
@RNKuhns, I don't think it makes sense for directional forecasters to return numbers - because that scientific type already has a name: forecaster.

I also wouln't be in favour of having a `predict_direction`, since the return may well depend on the aggregation, e.g., how we bin up/down/neutral, say.

The "sense" in differentiating directional forecasters from forecasters is that there are some directional forecasters which do *not* arise from binning a fitted forecaster. E.g., bin values, then fit a moving classifier with the original series being exogeneous and the binned discrete values being the series to forecast. This is a reduction step (= the first bullet point).

**Comment by RNKuhns** [comment]
@fkiraly thanks, that refreshes my memory. Makes sense with the generalized directional forecaster (moving classifier) that there is the need for the additional class. 

On the forecast evaluation side, I get the need for a `mean_directional_accuracy(y_true, y_pred)` that accepts `y_true` as real-valued and `y_pred` as a category so that we can evaluate the `DirectionalForecaster` class. 

But there are common use-cases of evaluating the directional accuracy of real valued forecasts (think GDP, unemployment rate or interest rate forecasts where there is a need to report a real valued number but still answer the question of how often the real valued prediction gets directional change correct). 

It would be nice if we could have the functionality for both cases. Any thoughts on how we could accommodate that?
